File: parser/src/main/java/com/github/fburato/highwheelmodules/bytecodeparser/FilteringDecorator.java
package com.github.fburato.highwheelmodules.model.bytecodeparser;

import com.github.fburato.highwheelmodules.model.classpath.AccessVisitor;
import com.github.fburato.highwheelmodules.model.bytecode.AccessPoint;
import com.github.fburato.highwheelmodules.model.bytecode.AccessType;
import com.github.fburato.highwheelmodules.model.bytecode.ElementName;

import java.util.function.Predicate;

/**
 * Passes calls through to wrapped child only if
 * they match the supplied filter
 */
class FilteringDecorator implements AccessVisitor {

  private final AccessVisitor child;
  private final Predicate<ElementName> filter;

  public FilteringDecorator(final AccessVisitor child, final Predicate<ElementName> filter) {
    this.child = child;
    this.filter = filter;
  }

  @Override
  public void apply(final AccessPoint source, final AccessPoint dest,
      final AccessType type) {
    if (this.filter.test(dest.getElementName()) && this.filter.test(source.getElementName())) {
      this.child.apply(source, dest, type);
    }
  }

  @Override
  public void newNode(final ElementName clazz) {
    if (this.filter.test(clazz)) {
      this.child.newNode(clazz);
    }
  }

  @Override
  public void newEntryPoint(ElementName clazz) {
    if (this.filter.test(clazz)) {
      this.child.newEntryPoint(clazz);
    }
  }

  @Override
  public void newAccessPoint(AccessPoint ap) {
    this.child.newAccessPoint(ap);
  }

}

File: parser/src/main/scala/com/github/fburato/highwheelmodules/bytecodeparser/FilteringDecorator.scala
package com.github.fburato.highwheelmodules.bytecodeparser

import com.github.fburato.highwheelmodules.model.bytecode.{AccessPoint, AccessType, ElementName}
import com.github.fburato.highwheelmodules.model.classpath.AccessVisitor

class FilteringDecorator(child: AccessVisitor, filter: ElementName => Boolean) extends AccessVisitor {

  override def apply(source: AccessPoint, dest: AccessPoint, `type`: AccessType): Unit = {
    if (filter(dest.elementName) && filter(source.elementName)) {
      child.apply(source, dest, `type`)
    }
  }

  override def newNode(clazz: ElementName): Unit = {
    if (filter(clazz)) {
      child.newNode(clazz)
    }
  }

  override def newAccessPoint(ap: AccessPoint): Unit = {
    child.newAccessPoint(ap)
  }

  override def newEntryPoint(clazz: ElementName): Unit = {
    if (filter(clazz)) {
      child.newEntryPoint(clazz)
    }
  }
}
